src/coreclr/src/jit/codegencommon.cpp

```diff
@@ -6308,71 +6308,165 @@ void CodeGen::genZeroInitFrame(int untrLclHi, int untrLclLo, regNumber initReg,
+            //    mov      rax, rdi
+            //    sub      rax, rcx    ; get unaligned amount
+            //    mov      ecx, <size> ; load amount to zero
+            //    sub      rcx, rax    ; subtract already zero'd unaligned
+            //    xor      rax, rax
+            //    rep stosb            ; zero the rest
+
+            noway_assert(regSet.rsRegsModified(RBM_EDI));
+#ifdef UNIX_AMD64_ABI
+            // For register arguments we may have to save ECX and RDI on Amd64 System V OSes
+            if (intRegState.rsCalleeRegArgMaskLiveIn & RBM_RCX)
+            {
+                noway_assert(regSet.rsRegsModified(RBM_R12));
+                inst_RV_RV(INS_mov, REG_R12, REG_RCX);
+                regSet.verifyRegUsed(REG_R12);
+            }
+
+            if (intRegState.rsCalleeRegArgMaskLiveIn & RBM_RDI)
+            {
+                noway_assert(regSet.rsRegsModified(RBM_R13));
+                inst_RV_RV(INS_mov, REG_R13, REG_RDI);
+                regSet.verifyRegUsed(REG_R13);
+            }
 #else  // !UNIX_AMD64_ABI
-        // For register arguments we may have to save ECX
-        if (intRegState.rsCalleeRegArgMaskLiveIn & RBM_ECX)
-        {
-            noway_assert(regSet.rsRegsModified(RBM_ESI));
-            inst_RV_RV(INS_mov, REG_ESI, REG_ECX);
-            regSet.verifyRegUsed(REG_ESI);
-        }
+            // For register arguments we may have to save ECX
+            if (intRegState.rsCalleeRegArgMaskLiveIn & RBM_ECX)
+            {
+                noway_assert(regSet.rsRegsModified(RBM_ESI));
+                inst_RV_RV(INS_mov, REG_ESI, REG_ECX);
+                regSet.verifyRegUsed(REG_ESI);
+            }
 #endif // !UNIX_AMD64_ABI
-
-        noway_assert((intRegState.rsCalleeRegArgMaskLiveIn & RBM_EAX) == 0);
-
-        GetEmitter()->emitIns_R_AR(INS_lea, EA_PTRSIZE, REG_EDI, genFramePointerReg(), untrLclLo);
-        regSet.verifyRegUsed(REG_EDI);
-
-        inst_RV_IV(INS_mov, REG_ECX, (untrLclHi - untrLclLo) / sizeof(int), EA_4BYTE);
-        instGen_Set_Reg_To_Zero(EA_PTRSIZE, REG_EAX);
-        instGen(INS_r_stosd);
+            // We need to 32 byte align the rep stosd as there is a penality if it is not 32 byte aligned
+            // Grab a non-argument, non-callee saved XMM reg
+            emitter*  emit     = GetEmitter();
+            regNumber frameReg = genFramePointerReg();
+            // Get block start
+            emit->emitIns_R_AR(INS_lea, EA_PTRSIZE, REG_EDI, frameReg, untrLclLo);
+#ifdef UNIX_AMD64_ABI
+            // System V x64 first temp reg is xmm8
+            regNumber zeroSIMDReg = genRegNumFromMask(RBM_XMM8);
+#else  // !UNIX_AMD64_ABI
+            // Windows first temp reg is xmm4
+            regNumber zeroSIMDReg = genRegNumFromMask(RBM_XMM4);
+#endif // UNIX_AMD64_ABI
+            // Zero xmm reg
+            emit->emitIns_R_R(INS_xorps, EA_16BYTE, zeroSIMDReg, zeroSIMDReg);
+            // Zero first 32 bytes (may overlap with next 32 byte alignment)
```

Comment:
As the stack is already 16 byte aligned, could we get away with just a single movdqu?

Reply:
For a 16 byte aligned stack; either way it will be a maximum overlap of 16 bytes and a minimum of 0 (as it always moves to next 32 byte boundary; rather than staying if it is on one)

This way is safer and simpler as if the stack is not perfectly 16 byte aligned (not necessarily in leap methods? https://github.com/dotnet/runtime/pull/32442#issuecomment-587165593 or on 32-bit) then it could leave gaps; so would need extra handling for them to fully align (as per this change).